experimental/Schemes/CoherentSheaves.jl

```diff
@@ -162,31 +163,373 @@ identifications given by the glueings in the `default_covering`.
       return MU
     end
 
+    function production_func(
+        F::AbsPreSheaf,
+        U::SimplifiedSpec
+      )
+      V = original(U)
+      MV = F(V)
+      rho = OOX(V, U)
+      MU, phi = change_base_ring(rho, MV)
+      add_incoming_restriction!(F, V, MU, phi)
+      return MU
+    end
+
     ### Production of the restriction maps; to be cached
     function restriction_func(F::AbsPreSheaf, V::AbsSpec, U::AbsSpec)
-      MV = F(V)
-      MU = F(U)
-      # There are two cases: Either U is a PrincipalOpenSubset of V, or U 
-      # is a PrincipalOpenSubset of another affine_chart W. In both cases, 
-      # we need to find W (which equals V in the first case) and use the transition 
-      # matrix for the changes between these two sets. 
-      VV = V
-      while VV isa PrincipalOpenSubset
-          VV = ambient_scheme(VV)
+      if any(W->(W === U), affine_charts(X)) && any(W->(W === V), affine_charts(X))
+        MV = F(V)
+        MU = F(U)
+        A = MG[(V, U)] # The transition matrix
+        UU, _ = glueing_domains(default_covering(X)[U, V])
+        psi = OOX(UU, U) # Needs to exist by the checks of is_open_func, even though 
+        # in general UU ⊂ U!
+        return hom(MV, MU, [sum([psi(A[i, j]) * MU[j] for j in 1:ngens(MU)]) for i in 1:ngens(MV)], rho)
+      else
+        error("invalid input")
+      end
+    end
+
+    function restriction_func(F::AbsPreSheaf, V::AbsSpec, U::PrincipalOpenSubset)
+      # If V was not an affine_chart of X, some other function would have 
+      # been triggered. 
+
+      # First the easy case: Inheritance from an ancestor in the tree.
+      if ambient_scheme(U) === V
+        # If the restriction was more complicated than what follows, then 
+        # it would have been cached earlier and this call would not have happened
```

Comment:
Relying on caching seems brittle to me. It is necessary?

Reply:
Why brittle?

First of all, it is default-stable in the sense that it doesn't produce wrong results in case nothing was cached. 

Second, the call to the `production_func` and `restriction_func` follows a clear protocol (not yet publicly documented, admittedly) which says that things are first looked up in the caches and only then produced if necessary. 

The idea is to use this with ideal sheaves or coherent sheaves which happen to be locally free: We want to ask for things like `locally_trivializing_cover` or `local_complete_intersection_cover`. These have to be implemented in a way so that 
  * the trivializing patches are `PrincipalOpenSubset`s or `SimplifiedSpec`s (or position themselves by other means in the tree-structure) with the results in the desired form on these patches
  * the restriction maps from their parent nodes are also in the prescribed form and, necessarily, cached.
 
Then, for every other patch `W` which sits below a trivializing patch `U`, we expect to also find restrictions which take into account the intermediate trivializations. That is also the reason why we go up recursively along the tree structure for the productions and restrictions (see @afkafkafk13 's comment above) and not straight to the top. 

Does this rule out your concerns?

Reply:
Brittle in the sense that a change of the caching behavior might affect this function and this could be unexpected. 
But if you say that it is stable by default, I am okay. 
For larger computations one might want to clear the caches from time to time / or allow them to be collected / or disable caching (where feasible) to save memory.

Reply:
Changing of the caching behaviour will certainly affect this function and whoever wants to do such things in obligated to also clean up here. 

Turning off caching is not an option here: We need the stability to have the very same object in memory whenever we call `M(U)` for a fixed `U` and a coherent sheaf `M` in order to maintain the element-parent structure in Oscar. 

I put a lot of effort into making the caching garbage collectable in the future. However, that still relies on introducing the `WeakKeyIdDict`s in Oscar at some point: https://github.com/JuliaCollections/DataStructures.jl/pull/402. I hope this will help with your concerns in this regard.